### PR TITLE
vimplugins: update.py supports GitHub Access tokens; add completion-nvim, completion-treesitter

### PR DESCRIPTION
**Pull request by colemickens**
###### Motivation for this change

Update vimplugins and add:
* https://github.com/haorenW1025/completion-nvim
* https://github.com/nvim-treesitter/completion-treesitter

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by teto — APPROVED** [review]
so last time I merged a neovim plugin for lsp, there were some complaints that we shouldn't package for an unreleased software. nvim 0.5 is still a few months away but works quite well already so I still believe it's fine to merge.

**Comment by colemickens** [comment]
I've rebased this and also made `./update.py` use `GITHUB_API_TOKEN` if defined. This makes updating considerably faster since you have use higher parallelism without 429s.

I think there are more than a few people using neovim 0.5, particularly for the features that these plugins revolve around. It certainly wouldn't be the first unstable thing packaged in nixpkgs, and at least one of them has a tagged release. Hopefully this can be merged. Thanks for considering/reviewing!

File: pkgs/misc/vim-plugins/generated.nix
Comment (teto) [inline pkgs/misc/vim-plugins/generated.nix]: a few plugins recently moved under the org "nvim-lua" so htis is the new home

**Comment by teto** [comment]
while at it would you mind packaging https://github.com/nvim-lua/lsp-status.nvim

**Comment by colemickens** [comment]
@teto Sure, I've redone the PR with `lsp-status.nvim` included as well.

Also, I was able to use `-p 16` to redo this, which was 16 times faster than can be achieved without `GITHUB_API_TOKEN`.

**Comment by colemickens** [comment]
this iteration isn't ready for merging, btw.

**Comment by colemickens** [comment]
Looks like I had missed another dependency of completion-treesitter.

@teto your requested plugin breaks my system config (it looks like maybe vimplugins tests that vim can startup properly?) When I point at this commit, it works. Maybe you can file an issue/pr on `lsp-status.nvim` for this? I don't quite have context, or time to test this outside of nixpkgs. https://github.com/colemickens/lsp-status.nvim/commit/3982991cdf72897ed047f5fd2197108751cb7dbe

I'm going to back out the `lsp-status-nvim` commit and push what works.

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:80]:
```suggestion
        hdr = {'Authorization': f'token {token}'}
        r = urllib.request.Request(url, headers=hdr)
        return r
    else:
        r = urllib.request.Request(url)
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:73]:
```suggestion
def makeRequest(url: str) -> urllib.request.Request:
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:102]:
```suggestion
            req = makeRequest(self.url(f"blob/{self.branch}/.gitmodules"))
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:114]:
```suggestion
        commit_req = makeRequest(commit_url)
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:81]:
```suggestion
    headers = {}
    if token is None:
       headers["Authorization"] = f'token {token}'
    return urllib.request.Request(url, headers=headers)
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:74]:
```suggestion
    token = os.getenv("GITHUB_API_TOKEN")
```

**Comment by colemickens** [comment]
Thanks @Mic92. I responded to feedback and ran `./update.py` again to validate and update.

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:73]: Actually I forgot we are doing python here. Did too much rust/go lately:

```suggestion
def make_request(url: str) -> urllib.request.Request:
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:99]:
```suggestion
            req = make_request(self.url(f"blob/{self.branch}/.gitmodules"))
```

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py:111]:
```suggestion
        commit_req = make_request(commit_url)
```

**Comment by Mic92** [comment]
Should the token be made a requirement? I mean without it, one does run into request limits quite fast, so it is not really usable.

**Comment by colemickens** [comment]
@Mic92 I don't know. I didn't want to make a "breaking change" if i I didn't have to. It *does* update with `-p 1`, it's just slow. I'd rather leave it to another PR, to be frank. (I went ahead and did this next round of fixups and rebased/pushed.)

**Review by Mic92 — APPROVED** [review]

**Comment by gvolpe** [comment]
Any chance you can update `coc-metals` as part of this PR as well? The current version in Nixpkgs is super outdated. The new repository is https://github.com/scalameta/coc-metals

I tried to update a few plugins but I was also blocked by https://github.com/NixOS/nixpkgs/issues/91339

**Comment by colemickens** [comment]
I don't think so, sorry. I've already churned this PR once for a change I wound up having to back out. I'd really like to see this merged so we can unblock everyone.

**Comment by gvolpe** [comment]
No problems, I'll try to put up a follow up PR.

**Comment by flokli** [comment]
Let's merge this :+1: Thanks for your work!

**Comment by gvolpe** [comment]
This still doesn't work for me. I have set up a `GITHUB_API_TOKEN` and some packages managed to be downloaded at first but then again Github doesn't seem to like the amount of requests. 

```
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
<urlopen error [Errno 113] No route to host>, Retrying in 3 seconds...
```

**Comment by colemickens** [comment]
Unless I made some other coding error, "No route to host" is a very different error condition than an HTTP 429.

**Comment by Mic92** [comment]
Agreed. Sounds like a local network issue.

**Comment by gvolpe** [comment]
I don't think I have any network issues, it would be nice if someone else could try and see if this works for them.

**Comment by colemickens** [comment]
I just ran it https://asciinema.org/a/mFnijt0sBehQyxZWUeRKb2Iu2 (and rotated the access token).

**Comment by gvolpe** [comment]
Wow, that really works for you! Thanks for that. Any chance you could try and make the same changes I made to `vim-plugin-names` [here](https://github.com/NixOS/nixpkgs/pull/92993) and then run the script again?

It's really late in my timezone but I promise to post what happens on my end tomorrow first thing. I wish this worked for me too...

**Comment by colemickens** [comment]
Sure, let's move the conversation over to your PR. Maybe a new issue would be best if you (or anyone) winds up finding something wrong with the update script.

**Pull request merged**
